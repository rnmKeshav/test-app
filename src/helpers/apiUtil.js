import request from "superagent";

/***********
options = {
  endpoint: '/get_users',
  query: {
    a: b
  },
  headers: {
    Content-Type: application/json
  },
  payload: {} // for POST/PATCH/DELETE
}
**********/
const apiRequest = (method, options = {}, host = "https://api.github.com/") => {
  let { endpoint, query = {}, payload = {} } = options;
  let headers = { accept: "application/json" };
  host = host || window.location.origin;
  let url = `${host}${endpoint}`;

  console.log("url", url);
  Object.assign(headers, options.headers);

  let req = request[method.toLowerCase()](url)
    .query(query)
    .set(headers);

  if (method.toLowerCase() !== "get") {
    req.send(payload);
  }

  return new Promise((resolve, reject) => {
    req.end((error, res) => {
      let response;

      if (error) {
        reject(error);
      }

      try {
        if (res && res.text !== "") {
          response = JSON.parse(res.text);
        }
      } catch (e) {
        reject(e);
      }

      resolve(response);
    });
  });
};

const api = {
  get: apiRequest.bind(null, "get"),
  post: apiRequest.bind(null, "post"),
  patch: apiRequest.bind(null, "patch"),
  delete: apiRequest.bind(null, "delete")
};

export default api;
