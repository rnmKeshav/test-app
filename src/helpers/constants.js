const actionTypes = {
  "LOGGEDIN_USER/DATA_FETCH_STARTED": "LOGGEDIN_USER/DATA_FETCH_STARTED",
  "LOGGEDIN_USER/DATA_FETCH_FAILURE": "LOGGEDIN_USER/DATA_FETCH_FAILURE",
  "LOGGEDIN_USER/DATA_FETCH_SUCCESS": "LOGGEDIN_USER/DATA_FETCH_SUCCESS",
  "LOGGEDIN_USER/CHANGE_HANDLE": "LOGGEDIN_USER/CHANGE_HANDLE",
  "LOGGEDIN_USER/UPDATE_TO_LOCALSTORE": "LOGGEDIN_USER/UPDATE_TO_LOCALSTORE"
};

module.exports = {
  actionTypes
};
