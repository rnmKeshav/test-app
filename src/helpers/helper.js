const debounce = (fn, delay) => {
  let timeout;

  return payload => (dispatch, getState) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => {
      dispatch(fn(payload));
    }, delay);
  };
};

module.exports = {
  debounce
};
