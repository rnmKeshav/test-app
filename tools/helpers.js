let parseArr = (arr) => (arg) => 
  arr
    .filter(item => !!item)
    .map(item => item(arg));

module.exports = {
  parseArr
};
