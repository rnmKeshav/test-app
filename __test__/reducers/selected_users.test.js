import selectedUsers from "../../src/reducers/selected_users";
import { actionTypes } from "../../src/helpers/constants";

describe("Selected_users reducer test", () => {
  let selectedUser = {
    login: "rnmp",
    id: 22627,
    node_id: "MDQ6VXNlcjIyNjI3",
    avatar_url: "https://avatars0.githubusercontent.com/u/22627?v=4",
    gravatar_id: "",
    url: "https://api.github.com/users/rnmp",
    html_url: "https://github.com/rnmp",
    followers_url: "https://api.github.com/users/rnmp/followers",
    following_url: "https://api.github.com/users/rnmp/following{/other_user}",
    gists_url: "https://api.github.com/users/rnmp/gists{/gist_id}",
    starred_url: "https://api.github.com/users/rnmp/starred{/owner}{/repo}",
    subscriptions_url: "https://api.github.com/users/rnmp/subscriptions",
    organizations_url: "https://api.github.com/users/rnmp/orgs",
    repos_url: "https://api.github.com/users/rnmp/repos",
    events_url: "https://api.github.com/users/rnmp/events{/privacy}",
    received_events_url: "https://api.github.com/users/rnmp/received_events",
    type: "User",
    site_admin: false,
    score: 32.814484
  };
  test("Should match initial state", () => {
    expect(
      selectedUsers(undefined, {
        type: actionTypes["SOME_UNDEFINED_ACTION"]
      })
    ).toMatchSnapshot();
  });

  test("Should add new user if not already exist", () => {
    expect(
      selectedUsers(undefined, {
        type: actionTypes["SELECTED_USER/ADD_NEW"],
        payload: {
          user: selectedUser
        }
      })
    ).toMatchSnapshot();
  });

  test("Should update user if already exist", () => {
    let existingUserState = selectedUsers(undefined, {
      type: actionTypes["SELECTED_USER/ADD_NEW"],
      payload: {
        user: selectedUser
      }
    });
    let updatedUser = Object.assign({}, selectedUser, { score: 34 });
    expect(
      selectedUsers(existingUserState, {
        type: actionTypes["SELECTED_USER/ADD_NEW"],
        payload: {
          user: updatedUser
        }
      })
    ).toMatchSnapshot();
  });
});
