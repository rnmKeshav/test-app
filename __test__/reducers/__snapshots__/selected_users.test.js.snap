// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selected_users reducer test Should add new user if not already exist 1`] = `
Object {
  "users": Array [
    Object {
      "avatar_url": "https://avatars0.githubusercontent.com/u/22627?v=4",
      "events_url": "https://api.github.com/users/rnmp/events{/privacy}",
      "followers_url": "https://api.github.com/users/rnmp/followers",
      "following_url": "https://api.github.com/users/rnmp/following{/other_user}",
      "gists_url": "https://api.github.com/users/rnmp/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/rnmp",
      "id": 22627,
      "login": "rnmp",
      "node_id": "MDQ6VXNlcjIyNjI3",
      "organizations_url": "https://api.github.com/users/rnmp/orgs",
      "received_events_url": "https://api.github.com/users/rnmp/received_events",
      "repos_url": "https://api.github.com/users/rnmp/repos",
      "score": 32.814484,
      "site_admin": false,
      "starred_url": "https://api.github.com/users/rnmp/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/rnmp/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/rnmp",
    },
  ],
}
`;

exports[`Selected_users reducer test Should match initial state 1`] = `
Object {
  "users": Array [],
}
`;

exports[`Selected_users reducer test Should update user if already exist 1`] = `
Object {
  "users": Array [
    Object {
      "avatar_url": "https://avatars0.githubusercontent.com/u/22627?v=4",
      "events_url": "https://api.github.com/users/rnmp/events{/privacy}",
      "followers_url": "https://api.github.com/users/rnmp/followers",
      "following_url": "https://api.github.com/users/rnmp/following{/other_user}",
      "gists_url": "https://api.github.com/users/rnmp/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/rnmp",
      "id": 22627,
      "login": "rnmp",
      "node_id": "MDQ6VXNlcjIyNjI3",
      "organizations_url": "https://api.github.com/users/rnmp/orgs",
      "received_events_url": "https://api.github.com/users/rnmp/received_events",
      "repos_url": "https://api.github.com/users/rnmp/repos",
      "score": 34,
      "site_admin": false,
      "starred_url": "https://api.github.com/users/rnmp/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/rnmp/subscriptions",
      "type": "User",
      "url": "https://api.github.com/users/rnmp",
    },
  ],
}
`;
